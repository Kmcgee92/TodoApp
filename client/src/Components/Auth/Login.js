import React, { useState, useEffect } from "react";
//redux
import { useDispatch, useSelector } from "react-redux";
import { loginHandler } from "../../redux/actions/authActions";
// apollo
import { useMutation } from "@apollo/react-hooks";
// import { GET_USER } from "../../graphql/queries/GetUser";
import { GET_USER_BY_LOGIN } from "../../graphql/mutations/login";

// mui components
import Button from "@material-ui/core/Button";
import Input from "@material-ui/core/Input";
import FormControl from "@material-ui/core/FormControl";
import FormHelperText from "@material-ui/core/FormHelperText";

const Login = ({ classes, setDataLoading, setModalOpen }) => {
  const dispatch = useDispatch();
  const auth = useSelector((state) => state.auth);
  const [email, setEmail] = useState("demo@bcf.com");
  const [password, setPassword] = useState("password");
  const [emailError, setEmailError] = useState(false);
  const [passwordError, setPasswordError] = useState(false);
  const [noUserFound, setNoUserFound] = useState(false);

  let [getUserByLogin, { data, loading, error }] = useMutation(
    GET_USER_BY_LOGIN
  );
  useEffect(() => {
    if (!error && data && !loading) {
      if (!data.Login) {
        setNoUserFound(true);
        return;
      }
      dispatch(loginHandler(data));
    }
  }, [dispatch, error, data, loading]);

  useEffect(() => {
    if (loading) {
      setDataLoading(true);
      //! setting timeout to make loader visible
      // directly based off query for user info loading param
      setTimeout(() => setDataLoading(false), 2800);
    }
  }, [loading, setDataLoading]);

  // error handling for login and perform DB Query which should be done in redux thunk
  const handleLogin =  () => {
    setModalOpen(false);
    setNoUserFound(false);
    setEmailError(false);
    setPasswordError(false);
    let emailCurrentError = false;
    let passwordCurrentError = false;

    if (!email) {
      emailCurrentError = true;
      setEmailError(true);
    }

    if (!password) {
      passwordCurrentError = true;
      setPasswordError(true);
    }

    const regex = /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;

    if (!regex.test(email)) {
      emailCurrentError = true;
      setEmailError(true);
    }

    if (!emailCurrentError && !passwordCurrentError) {
      setDataLoading(true);
      getUserByLogin({
        variables: {
          email: email,
          password: password,
        },
      });
    }
    setDataLoading(false);
  };

  return (
    <div>
      <form className={classes.signinForm}>
        {auth.error && (
          <FormHelperText className={classes.serverError}>
            {`${auth.error}`}
          </FormHelperText>
        )}
        {(emailError || passwordError) && (
          <FormHelperText style={{ color: "red" }}>
            please fill out all fields.
          </FormHelperText>
        )}
        {noUserFound && (
          <FormHelperText style={{ color: "red" }}>
            500: Seems to be an issue with the server
          </FormHelperText>
        )}
        <FormControl>
          <Input
            required
            data-test="loginEmail"
            value={email}
            className={classes.authInputs}
            placeholder="email"
            autoComplete="email"
            autoFocus={true}
            error={emailError}
            onChange={(e) => setEmail(e.target.value)}
            onClick={() => setModalOpen(false)}
            />
        </FormControl>
        <FormControl>
          <Input
            required
            data-test="loginPassword"
            value={password}
            type="password"
            className={classes.authInputs}
            placeholder="password"
            autoComplete="current-password"
            error={passwordError}
            onChange={(e) => setPassword(e.target.value)}
            onClick={() => setModalOpen(false)}
          />
        </FormControl>
        <Button
          data-test="loginButton"
          onClick={handleLogin}
          className={classes.icon}
        >
          Login
        </Button>
      </form>
    </div>
  );
};

export default Login;