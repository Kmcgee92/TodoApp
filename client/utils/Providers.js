import configureStore from "redux-mock-store";
import thunk from "redux-thunk";
// provide redux and apollo
import { Provider } from "react-redux";
import { MockedProvider } from "@apollo/client/testing";

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

export default function ComponentWithProviders(
  ComponentToBeTested,
  props,
  initialState
) {
  let mocks = [];
  let store = mockStore(
    
    initialState || { auth: { activeUser: {}, error: false } }
  
  );
  return (
    <MockedProvider mocks={mocks}>
      <Provider store={store}>
        <ComponentToBeTested {...props} />
      </Provider>
    </MockedProvider>
  );
}
