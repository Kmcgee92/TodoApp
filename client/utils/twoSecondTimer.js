export default function twoSecondTmer() {
  return new Promise((resolve) => {
    setTimeout(resolve, 2000);
  });
}
