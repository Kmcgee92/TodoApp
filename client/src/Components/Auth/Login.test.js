import React from "react";
import { act } from "react-dom/test-utils";

import { shallow, render, mount } from "enzyme";
// component to test
import Login from "./Login";
// child components
import Input from "@material-ui/core/Input";
import Button from "@material-ui/core/Button";

// utils
import Providers from "../../../utils/Providers";
import { findByTestAtrr } from "../../../utils/index";

describe("<Login/>", () => {
  let props;
  let wrapWithProviders;
  let component;
  describe("Component Initial Rendering", () => {
    beforeEach(() => {
      props = {
        classes: {},
      };
      wrapWithProviders = mount(Providers(Login, props));
      component = wrapWithProviders.find(Login);
    });
    it("Should Render", () => {
      expect(component).toHaveLength(1);
    });

    it("Should Match Snapshot", () => {
      wrapWithProviders = mount(Providers(Login, props));
      component = wrapWithProviders.find(Login);
      expect(component).toMatchSnapshot();
    });
  });

  describe("Rendering Children", () => {
    beforeEach(() => {
      props = {
        classes: {},
      };
      wrapWithProviders = mount(Providers(Login, props));
      component = wrapWithProviders.find(Login);
    });
    it("Should Render 'form' Element", () => {
      expect(component.find("form")).toHaveLength(1);
    });
    it("Should Render Two Inputs on Mount", () => {
      expect(component.find("input")).toHaveLength(2);
    });
    it("Should Render Login Button on Mount", () => {
      let button = component.find(Button);
      expect(button).toBeDefined();
    });
  });

  describe("Form Input Values Are Updateable and Submittable", () => {
    beforeEach(() => {
      props = {
        classes: {},
        setDataLoading: jest.fn(),
        setModalOpen: jest.fn(),
      };
      wrapWithProviders = mount(Providers(Login, props));
      component = wrapWithProviders.find(Login);
    });
    it("Should Capture Email correctly onChange", () => {
      let emailWrapper = findByTestAtrr(component, "loginEmail").at(4);
      let emailInput = emailWrapper.find("input");
      expect(emailInput.props().value).toEqual("demo@bcf.com");
      emailInput.props().value = "mockEmail";
      expect(emailInput.props().value).toEqual("mockEmail");
    });
    it("Should Capture Password correctly onChange", () => {
      let passwordWrapper = findByTestAtrr(
        wrapWithProviders,
        "loginPassword"
      ).at(4);
      let passwordInput = passwordWrapper.find("input");
      expect(passwordInput.props().value).toEqual("password");
      passwordInput.props().value = "mockPassword";
      expect(passwordInput.props().value).toEqual("mockPassword");
    });
    it("When Form is Submitted, handleLogin is Called", () => {
      let button = component.find(Button);
      expect(button).toBeDefined();
      button.simulate("click");
      expect(props.setDataLoading).toBeCalledTimes(3);
      expect(props.setModalOpen).toBeCalledTimes(1);
    });
  });
});
