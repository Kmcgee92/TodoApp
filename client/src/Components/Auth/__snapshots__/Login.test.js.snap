// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Login/> Component Initial Rendering Should Match Snapshot 1`] = `ReactWrapper {}`;
